### ISSUE TITLE
open pidfd: no such process

### ISSUE THREAD

**Issue by edsantiago** [containers/podman#18452]
Not a new flake -- this one started cropping up in 2020, but never often enough to even merit an issue. Now, it's happening constantly in #18442, a PR that adds error checking to e2e-test cleanup.
```
$ podman [options] stop --all -t 0
open pidfd: No such process
```
This may not be a complete list--it has been overwhelming:

* fedora-37 : int podman fedora-37 root container boltdb
    * [05-03 11:21](https://api.cirrus-ci.com/v1/artifact/task/5506136472813568/html/int-podman-fedora-37-root-container-boltdb.log.html#t--podman-pod-container-can-override-pod-not-sharing-pid--1) in podman pod container can override pod not sharing pid
* fedora-38 : int podman fedora-38 root container boltdb
    * [05-03 11:17](https://api.cirrus-ci.com/v1/artifact/task/6069086426234880/html/int-podman-fedora-38-root-container-boltdb.log.html#t--podman-pod-correctly-sets-up-PIDNS--1) in podman pod correctly sets up PIDNS
* fedora-38 : int podman fedora-38 root host boltdb
    * [05-03 11:19](https://api.cirrus-ci.com/v1/artifact/task/6209823914590208/html/int-podman-fedora-38-root-host-boltdb.log.html#t--podman-generate-kube-sharing-pid-namespace--1) in podman generate kube sharing pid namespace
* fedora-38 : int podman fedora-38 root host sqlite
    * [05-03 11:19](https://api.cirrus-ci.com/v1/artifact/task/4802449031036928/html/int-podman-fedora-38-root-host-sqlite.log.html#t--podman-play-kube-should-share-ipc-net-uts-when-shareProcessNamespace-is-set--1) in podman play kube should share ipc,net,uts when shareProcessNamespace is set
    * [05-03 11:19](https://api.cirrus-ci.com/v1/artifact/task/4802449031036928/html/int-podman-fedora-38-root-host-sqlite.log.html#t--podman-play-kube-test-with-init-containers-and-annotation-set--1) in podman play kube test with init containers and annotation set
    * [05-03 11:19](https://api.cirrus-ci.com/v1/artifact/task/4802449031036928/html/int-podman-fedora-38-root-host-sqlite.log.html#t--podman-play-kube-multi-doc-yaml-with-multiple-services--pods-and-deployments--1) in podman play kube multi doc yaml with multiple services, pods and deployments
    * [05-03 11:19](https://api.cirrus-ci.com/v1/artifact/task/4802449031036928/html/int-podman-fedora-38-root-host-sqlite.log.html#t--podman-play-kube-with-replicas-limits-the-count-to-1-and-emits-a-warning--1) in podman play kube with replicas limits the count to 1 and emits a warning
    * [05-03 11:19](https://api.cirrus-ci.com/v1/artifact/task/4802449031036928/html/int-podman-fedora-38-root-host-sqlite.log.html#t--podman-play-kube-override-with-udp-should-keep-tcp-from-YAML-file--1) in podman play kube override with udp should keep tcp from YAML file
* fedora-38 : int podman fedora-38 rootless host sqlite
    * [05-03 11:18](https://api.cirrus-ci.com/v1/artifact/task/4591342798503936/html/int-podman-fedora-38-rootless-host-sqlite.log.html#t--podman-play-kube-test-with-init-containers-and-annotation-set--1) in podman play kube test with init containers and annotation set
    * [05-03 11:18](https://api.cirrus-ci.com/v1/artifact/task/4591342798503936/html/int-podman-fedora-38-rootless-host-sqlite.log.html#t--podman-play-kube-deployment-more-than-1-replica-test-correct-command--1) in podman play kube deployment more than 1 replica test correct command
    * [05-03 11:18](https://api.cirrus-ci.com/v1/artifact/task/4591342798503936/html/int-podman-fedora-38-rootless-host-sqlite.log.html#t--podman-play-kube-multi-doc-yaml-with-multiple-services--pods-and-deployments--1) in podman play kube multi doc yaml with multiple services, pods and deployments
* rawhide : int podman rawhide root host sqlite
    * [05-03 11:18](https://api.cirrus-ci.com/v1/artifact/task/5365398984458240/html/int-podman-rawhide-root-host-sqlite.log.html#t--podman-generate-kube-with-volume--1) in podman generate kube with volume
    * [05-03 11:18](https://api.cirrus-ci.com/v1/artifact/task/5365398984458240/html/int-podman-rawhide-root-host-sqlite.log.html#t--podman-generate-kube-sharing-pid-namespace--1) in podman generate kube sharing pid namespace
    * [05-03 11:18](https://api.cirrus-ci.com/v1/artifact/task/5365398984458240/html/int-podman-rawhide-root-host-sqlite.log.html#t--podman-play-kube-should-share-ipc-net-uts-when-shareProcessNamespace-is-set--1) in podman play kube should share ipc,net,uts when shareProcessNamespace is set
    * [05-03 11:18](https://api.cirrus-ci.com/v1/artifact/task/5365398984458240/html/int-podman-rawhide-root-host-sqlite.log.html#t--podman-play-kube-expose-character-device-inside-container--1) in podman play kube expose character device inside container
    * [05-03 11:18](https://api.cirrus-ci.com/v1/artifact/task/5365398984458240/html/int-podman-rawhide-root-host-sqlite.log.html#t--podman-play-kube-override-with-udp-should-keep-tcp-from-YAML-file--1) in podman play kube override with udp should keep tcp from YAML file
    * [05-03 11:18](https://api.cirrus-ci.com/v1/artifact/task/5365398984458240/html/int-podman-rawhide-root-host-sqlite.log.html#t--podman-pod-correctly-sets-up-PIDNS--1) in podman pod correctly sets up PIDNS
* rawhide : int podman rawhide rootless host sqlite
    * [05-03 11:17](https://api.cirrus-ci.com/v1/artifact/task/5154292751925248/html/int-podman-rawhide-rootless-host-sqlite.log.html#t--podman-play-kube-support-container-liveness-probe--1) in podman play kube support container liveness probe
    * [05-03 11:17](https://api.cirrus-ci.com/v1/artifact/task/5154292751925248/html/int-podman-rawhide-rootless-host-sqlite.log.html#t--podman-play-kube-test-with-init-containers-and-annotation-set--1) in podman play kube test with init containers and annotation set
    * [05-03 11:17](https://api.cirrus-ci.com/v1/artifact/task/5154292751925248/html/int-podman-rawhide-rootless-host-sqlite.log.html#t--podman-play-kube-test-with-DirectoryOrCreate-HostPath-type-volume-and-non-existent-directory-path--1) in podman play kube test with DirectoryOrCreate HostPath type volume and non-existent directory path
    * [05-03 11:17](https://api.cirrus-ci.com/v1/artifact/task/5154292751925248/html/int-podman-rawhide-rootless-host-sqlite.log.html#t--podman-generate-and-reimport-kube-on-pod--1) in podman generate and reimport kube on pod

**Comment by vrothberg** [comment]
I am under the impression that's `crun` error (see https://github.com/containers/crun/blob/645b75b04613debb0a3bdf122fdb0b22d835db68/src/libcrun/linux.c#L5320).

@giuseppe, @flouthoc, does `crun` need some error handling as done in Podman:
```golang
// Track lifetime of conmon precisely using pidfd_open + poll.           
// There are many cases for this to fail, for instance conmon is dead    
// or pidfd_open is not supported (pre linux 5.3), so fall back to the   
// traditional loop with poll + sleep                                    
if fd, err := unix.PidfdOpen(c.state.ConmonPID, 0); err == nil {         
        return fd                                                        
} else if err != unix.ENOSYS && err != unix.ESRCH {                      
        logrus.Debugf("PidfdOpen(%d) failed: %v", c.state.ConmonPID, err)
}                                                                        
```

**Comment by giuseppe** [comment]
crun must report when the kill command fails, as it is in this case because the process doesn't exist.  Should Podman ignore it?

**Comment by vrothberg** [comment]
> crun must report when the kill command fails, as it is in this case because the process doesn't exist. Should Podman ignore it?

That sounds good to me.  Could `crun` return with a specific exit code such that Podman can discriminate a no-such-pid error from others?  I'd like to avoid parsing error strings.

**Comment by vrothberg** [comment]
Meh ... but `runc` and other runtimes wouldn't behave that way.

**Comment by vrothberg** [comment]
I have an idea.  Let me wrap something up to discuss the details in the PR.

**Comment by vrothberg** [comment]
-> https://github.com/containers/podman/pull/18457

**Comment by Luap99** [comment]
Maybe a different angle, all of the above tests use kube or pods directly.  So it could be very well a bug in our pod handling.

**Comment by vrothberg** [comment]
> Maybe a different angle, all of the above tests use kube or pods directly. So it could be very well a bug in our pod handling.

Nice idea.  Let's see if the logs pop up and then we can continue digging.

**Comment by edsantiago** [comment]
Seen in my logs on [July 25](https://api.cirrus-ci.com/v1/artifact/task/6732338579636224/html/int-podman-ubuntu-2110-rootless-host.log.html#t--podman-generate-kube-sharing-pid-namespace--1) but I need to reiterate my usual warning about this sort of flake: it does not cause a CI failure, even in my no-retries PR, so the only way for me to see it is if it happens in the same logfile as a flake that I _do_ detect. IOW, this is likely happening often with no indication.

**Comment by Luap99** [comment]
If you want to make this fatal pick https://github.com/containers/podman/pull/18442 in your no retry PR.

**Comment by edsantiago** [comment]
I considered it but didn't see the point? As I recall, this was (is?) such a frequent one that it would be exhausting to sift through; like the "no logs from conmon" one. 

But I'll give it a try, gather some data, and report back.

**Comment by edsantiago** [comment]
Oh.... [now](https://api.cirrus-ci.com/v1/artifact/task/4734999807655936/html/int-podman-fedora-37-root-host-sqlite.log.html) - [I](https://api.cirrus-ci.com/v1/artifact/task/4559077947211776/html/int-podman-fedora-38-root-container-sqlite.log.html) - [remember](https://api.cirrus-ci.com/v1/artifact/task/5297949761077248/html/int-podman-debian-13-root-host-sqlite.log.html) - [why](https://api.cirrus-ci.com/v1/artifact/task/6564587156275200/html/int-podman-fedora-38-root-host-sqlite.log.html#t--Podman-play-kube-podman-kube-play-test-with-reserved-CIDFile-annotation-in-yaml--1) - [I](https://api.cirrus-ci.com/v1/artifact/task/5438687249432576/html/int-podman-fedora-38-root-host-boltdb.log.html#t--Podman-play-kube-podman-play-kube-test-with-hostPID--1) - [stopped](https://api.cirrus-ci.com/v1/artifact/task/5438687249432576/html/int-podman-fedora-38-root-host-boltdb.log.html#t--Podman-play-kube-podman-play-kube-test-with-hostPID--1) - [doing](https://api.cirrus-ci.com/v1/artifact/task/5403502877343744/html/int-podman-debian-13-rootless-host-sqlite.log.html#t--Podman-pod-create-podman-pod-container-can-override-pod-not-sharing-pid--1) - [this](https://api.cirrus-ci.com/v1/artifact/task/5684977854054400/html/int-podman-fedora-37-root-container-sqlite.log.html).

**Comment by edsantiago** [comment]
Agreed! But, like my no-flake-retries work, it can only succeed if we make a concerted effort to squash all bugs and then enable the checks. And until that happens, new failures are going to creep in.

FWIW the `kube` failure is a constant, not a flake, so it might be easy to fix. At first glance maybe `kube down` could fix it. I'll try to take a look today or tomorrow.

**Comment by edsantiago** [comment]
Seen just now on my laptop, running simple interactive commands, not in int tests or anything:
```console
$ time bin/podman kube play --service-container /tmp/foo.yaml
...
$ bin/podman kube down /tmp/foo.yaml
2023-08-23T13:39:24.557276Z: open pidfd: No such process
Pods stopped:
...
```

**Comment by rhatdan** [comment]
@giuseppe This is an error returned from crun meaning the pid no longer exists.  It is called in the step where we are killing the pidfd,  Should we ignore this error?

**Comment by edsantiago** [comment]
I think we _do_ ignore the error. At least, podman does not pass it on in the exit code. 

The problem is that the message is shown to the user. This (1) causes tests to fail if they're expecting empty output, and (2) causes concern in end users, probably also wasted time (what happened? Are my containers ok? I need to look; Maybe I should check the web).

**Comment by rhatdan** [comment]
Well if the message is being shown, isn't it coming from crun then?  Podman is probably ignoring the error, but someone is writing it to stdout/stderr.

**Comment by giuseppe** [comment]
> @giuseppe This is an error returned from crun meaning the pid no longer exists. It is called in the step where we are killing the pidfd, Should we ignore this error?

yes I think we should ignore it if we try to kill the process.  It means the container already exited

**Comment by giuseppe** [comment]
tentative fix: https://github.com/containers/podman/pull/19760

not 100% sure it solves the problem as I was not able to reproduce locally yet to confirm it
